Jarl post (#183)
r-hub/blog: content/post/2026-06-01-jarl/index.Rmd

---
slug: jarl
title: "Refactoring with Jarl: a coffee chat" 
authors: 
- Hannah Frick
- Maëlle Salmon
date: "2026-06-01" 
tags: 
- package development
- programming
output: hugodown::hugo_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.path = "", comment = "")
# knitr hook to make images output use Hugo options
knitr::knit_hooks$set(
  plot = function(x, options) {
    hugoopts <- options$hugoopts
    paste0(
      "{{<figure src=",
      '"',
      x,
      '" ',
      if (!is.null(hugoopts)) {
        glue::glue_collapse(
          glue::glue('{names(hugoopts)}="{hugoopts}"'),
          sep = " "
        )
      },
      ">}}\n"
    )
  }
)
```

```{r, include = FALSE, cache = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  out.width = "100%",
  asciicast_theme = "pkgdown"
)
```

```{=html}
{{< figure src="meme.png" alt="Cute kitten attacked by robots. The text says 'Everytime you use Claude for something a CLI can do, a kitten dies'." >}}

```

> TL;DR, don’t let your friends use Claude for finding useless code in a project! 

Reviewer: Do we want to replace Claude with generally LLMs here? I don't have a real preference, more wondering out loud.

Using [Jarl](https://jarl.etiennebacher.com/) instead is cheaper, more reliable, and won’t kill any kitten.

We (Hannah and Maëlle) share an appreciation for the unglamorous maintenance work we call upkeep. 
So when Claude highlighted some [dead](https://github.com/tidymodels/tune/commit/5b4d63b58bd496e6c2809a5444b6ab119ae14f37) [code](https://github.com/tidymodels/tune/commit/16766b32ae1157425b124a70ec0bd64246e24c7f) in the tune package to Hannah, it was worth a mention during a chat. 
This led to Maëlle recommending Etienne Bacher's [Jarl](https://jarl.etiennebacher.com/), a fast linter for R written in Rust that can, among other things, detect unused functions.

Motivated by this conversation among other things, Maëlle wrote a [whole blog post](https://ropensci.org/blog/2026/04/02/tree-sitter-overview/) about Jarl and other tooling for R based on tree-sitter, including the above meme.

So naturally, when you get called out in a custom meme, you listen. :smile_cat:

> So here’s the long(er) form of “Use Jarl!”, from my following sweep through the [parsnip](https://parsnip.tidymodels.org/) package, with commentary from Maëlle.

Reviewer: ```suggestion
So here’s the long(er) form of “Use Jarl!”, from the following sweep through the [parsnip](https://parsnip.tidymodels.org/) package by Hannah, with commentary from Maëlle.
```
is this better?

I could have run ` jarl check . --fix` to fix everything with an automatic fix but I wanted to learn a bit more about the rules to see if I wanted to apply them. 
So I worked my way through one rule at a time, with one git commit at a time. 
You can [select rules](https://jarl.etiennebacher.com/by-example#select-or-ignore-specific-rules) like so:

```sh
jarl check . --select any_is_na
```

This separation of changes by rule also made it easier to review changes: one topic per commit.
_Maëlle: in igraph, I created one **PR** per rule or group of rules._ 

I first went through the ones with automatic fixes, then the ones without.
The result of the clean-up: https://github.com/tidymodels/parsnip/pull/1356
_Maëlle: in igraph, I went through rules starting with the ones with the least violations, ending with the ones that had the most hits._

parsnip has been around the block a few times and had accumulated a few functions flagged by Jarl as unused. 
I spotted `release_bullets()` in the list which I know we want to keep.[^release]

> We use it for parsnip in the release process but we don't use it in parsnip. 

Reviewer: ```suggestion
We use it _for_ parsnip in the release process but we don't use it _in_ parsnip. 
```

I wasn't expecting Jarl to catch onto this difference so I decided to review the list of flagged functions separately and split that off into its own issue (and PR). 

> I ended up keeping one other function that was part of a set of functions provided in a standalone file but the other ones I removed. 

Reviewer: ```suggestion
I ended up keeping one other function that was part of a set of functions provided in a standalone file but I removed the other ones. 
```

[^release]: This inspired an [issue](https://github.com/etiennebacher/jarl/issues/497).

No kittens were hurt!

## Why use Jarl for refactoring

Obviously, because we are telling you you should!
More seriously: Jarl, a command line interface (CLI) for linting code, is fast and free!

And here are a few more reasons beyond our success stories :grin::

### Jarl will find small smelly things

For instance, the [`outer_negation`](https://jarl.etiennebacher.com/rules/outer_negation ) rule helps improve readability: `!all(x)` is easier to understand than `any(!x)`.
See it in action in [parsnip](https://github.com/tidymodels/parsnip/pull/1356/changes/8029bb8eb9acab8beef75975fd6675e38a70f802).

> Generally, Jarl will make the codebase less [smelly](https://github.com/jennybc/code-smells-and-feels and https://www.youtube.com/watch?v=7oyiPBjLAWY ).

Reviewer: Does this linking syntax work? 👀

### Jarl will detect unreachable code

For instance, code that comes after a `return()` or a `stop()` in a function.
Sometimes, you can simply delete that code (again, a feel good move!).
Other times, the check is an opportunity for [targeted refactoring](https://github.com/tidymodels/parsnip/issues/1359)!

## Conclusion

Automatic tools are extremely useful for guiding upkeep work. 
Thinking _“let me clean up this repo”_ can be daunting :scream:, tools like Jarl provide a roadmap!
They let you fix obvious problems, and at the same time, they can make you read some dusty corners of your codebase.
Hannah's work on parsnip inspired Maëlle to run Jarl on igraph (influencer influenced back!), and she too can confirm it was both useful and satisfying.
Thanks Etienne for that cool tool!

Note that you could run Jarl checks manually as we did, or you could tell a LLM to run Jarl. 
To quote Hannah's colleague [Emil Hvitfeldt](https://emilhvitfeldt.com/post/ast-grep-r-claude-code/ ), _“When coding or using coding agents, one way to improve your workflow is by using CLI tools with a very clear focus.”_. 🫡

What's even better with Jarl is that it is actively developed!

> The [open PR](https://github.com/etiennebacher/jarl/pull/454) to add a check for unused _function arguments_ is in particular very exciting! 

Reviewer: ```suggestion
The [open PR](https://github.com/etiennebacher/jarl/pull/454) to add a check for unused _function arguments_ is very exciting in particular! 
```

Author: @hfrick did you mean to put the adjective right after the verb?

Reviewer: Gah 😆 yes. I meant "is very exciting in particular". I don't have a reference for why I would put "very exciting" before "in particular" but it "feels" like the right order to me.
